cylc/uiserver/app.py

```diff
@@ -0,0 +1,369 @@
+#!/usr/bin/env python3
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+from pathlib import Path, PurePath
+import sys
+from typing import List
+
+from pkg_resources import parse_version
+from tornado import ioloop
+from tornado.web import RedirectHandler
+from traitlets import (
+    Float,
+    TraitError,
+    TraitType,
+    Undefined,
+    Unicode,
+    default,
+    validate,
+)
+
+from jupyter_server.base.handlers import JupyterHandler
+from jupyter_server.extension.application import ExtensionApp
+# from jupyterhub.utils import url_path_join
+import tornado
+
+from cylc.flow.cfgspec.globalcfg import GlobalConfig
+from cylc.flow.network.graphql import (
+    CylcGraphQLBackend, IgnoreFieldMiddleware
+)
+
+from cylc.uiserver import (
+    __file__ as uis_pkg
+)
+from cylc.uiserver.data_store_mgr import DataStoreMgr
+from cylc.uiserver.handlers import (
+    SubscriptionHandler,
+    UIServerGraphQLHandler,
+    UserProfileHandler,
+    CylcStaticHandler
+)
+from cylc.uiserver.resolvers import Resolvers
+from cylc.uiserver.schema import schema
+from cylc.uiserver.websockets.tornado import TornadoSubscriptionServer
+from cylc.uiserver.workflows_mgr import WorkflowsManager
+
+
+class MyExtensionHandler(JupyterHandler):
+
+    @tornado.web.authenticated
+    def get(self):
+        self.write('hello cylc')
+
+    @tornado.web.authenticated
+    def post(self):
+        ...
+
+
+class PathType(TraitType):
+    """A pathlib traitlet type which allows string and undefined values."""
+
+    @property
+    def info_text(self):
+        return 'a pathlib.PurePath object'
+
+    def validate(self, obj, value):
+        if isinstance(value, str):
+            return Path(value).expanduser()
+        if isinstance(value, PurePath):
+            return value
+        if value == Undefined:
+            return value
+        self.error(obj, value)
+
+
+class CylcUIServer(ExtensionApp):
+
+    name = 'cylc'
+    app_name = 'cylc-gui'
+    default_url = "/cylc"
+    load_other_extensions = True
+    # file_url_prefix = "/render"
+    description = '''
+    Cylc - A user interface for monitoring and controlling Cylc workflows.
+    '''
+    examples = '''
+        cylc gui    # start the cylc GUI
```

Comment:
Note: I think there's a NBSP character here. Even though GH UI doesn't show it, my IDE distinguishes spaces and NBSP's.

![image](https://user-images.githubusercontent.com/304786/129639880-f98f7e16-8d96-4ee2-b306-b11c959ec818.png)

Harmless, but may break some tools/editors (I think moving cursor over the space and pressing `ga` in vim shows the hexcode/etc to compare with space)

Reply:
Harmless but can be annoying, it's what happens when you press alt+space on Mac OS.